src/oneid/jwts.py

```
@@ -0,0 +1,373 @@
+            'kid': keypair.identity,
+        }
+        header.update(MINIMAL_JSON_JWS_HEADER)
+        header_b64 = utils.to_string(utils.base64url_encode(json.dumps(header)))
+        to_sign = '{header}.{claims}'.format(header=header_b64, claims=payload)
+
+        signature = utils.to_string(keypair.sign(to_sign))
+
+        ret['signatures'].append({
+            'protected': header_b64,
+            'header': header,
+            'signature': signature,
+        })
+
+    return json.dumps(ret)
+
+
+def verify_jws(jws, keypairs=None, verify_all=True):
+    """
+    Convert a JWS back to it's claims, if validated by a set of
+    required :py:class:`~oneid.keychain.Keypair`s
+
+    :param jws: JWS to verify and convert
+    :type jws: dict, str or bytes
+    :param keypairs: :py:class:`~oneid.keychain.Keypair`s to verify the JWS with.
+                    Must include one for each specified in the JWS headers' `kid` values.
+    :type keypairs: list of :py:class:`~oneid.keychain.Keypair`
+    :param verify_all: If True (default), all keypairs must validate a signature.
+                    If False, only one needs to.
+                    If any fail to validate, the JWS is still not validated.
+                    This allows the caller to send multiple keys that _might_ have
+                    corresponding signatures, without requiring that _all_ do.
+    :type verify_all: bool
+    """
+
+    if not isinstance(jws, dict):
+        jws = utils.to_string(jws)
+
+        if re.match(COMPACT_JWS_RE, jws):
+
+            if verify_all and keypairs and len(keypairs) != 1:
+                raise Exception('Compact JWS (JWT) found but multiple signatures required')
+
+            return verify_jwt(jws, keypairs and keypairs[0])
+        jws = json.loads(jws)
+
+    claims = _verify_claims(utils.to_string(utils.base64url_decode(jws['payload'])))
+
+    if claims is None:
+        logger.debug('invalid claims: %s', jws['payload'])
+        return False
```

If verification fails, it should raise an exception.